llmfoundry/data/packing.py

```diff
@@ -360,9 +360,13 @@ def build_dataloader(cfg: DictConfig, tokenizer: PreTrainedTokenizerBase,
     # build tokenizer
     if 'tokenizer' not in cfg:
         raise ValueError('config must define tokenizer')
-    tokenizer_cfg: Dict[str,
-                        Any] = om.to_container(cfg.tokenizer,
-                                               resolve=True)  # type: ignore
+
+    resolved_tokenizer_cfg = om.to_container(cfg.tokenizer, resolve=True)
```

Comment:
The return types of `om.to_container` cause many issues, and is the source of many other unfixed `type: ignore` in the codebase

I'm unsure how to solve the type coercion more broadly, are there cases where the provided `cfg` could cause a return of a `List, Any, str`, such that we legitimately need a type guard here?

Reply:
Ah, that's unfortunate that there's no type narrowing...What if you cast to a `Dict[str, Any]`? Or you could assert that tokenizer_cfg is a `Dict`?

Reply:
Another thought: It seems like most of the usages are for building a tokenizer. Perhaps a shared utility function to create a tokenizer from configs is called for? That way, we could do assertion and casting all in a single place without bloating the codebase. 

I'm thinking something like:
```
def tokenizer_from_cfg(cfg: DictConfig) -> Dict[str, Any]:
    # build tokenizer
    if 'tokenizer' not in cfg:
        raise ValueError('config must define tokenizer')
    tokenizer_cfg = om.to_container(cfg.tokenizer, resolve=True)
    assert isinstance(tokenizer_cfg, Dict)
    tokenizer_name = tokenizer_cfg['name']
    tokenizer_kwargs = tokenizer_cfg.get('kwargs', {})
    tokenizer = build_tokenizer(tokenizer_name, tokenizer_kwargs)
    return tokenizer

# other stuff...

assert isinstance(DictConfig, cfg)
tokenizer_from_cfg(cfg)
```

Reply:
What about the way that chuck handled everything in `train.py`? https://github.com/mosaicml/llm-foundry/blob/dc1374816a007488ca9885c92e83e89e6fe933ef/scripts/train/train.py#L220-L234

Reply:
oh whoa, how does that type check correctly if pop_config returns `Any` ?

Reply:
I think you're telling it that you _know_ the types don't match, but you're doing it anyway. similar to `assert isinstance` but without the assert part? I think this is probably ok for this usage...

Reply:
I could use `pop_config`... it feels purely performative though, since that method doesn't actually enforce `isinstance(cfg.tokenizer, DictConfig)` -- I might as well have just done `cast(Dict, om.to_container(...))` and removed a level of indirection. Perhaps we can huddle quickly on this tomorrow.

Reply:
Yeah, agreed its performative. I've just asserted dictconfig in other places and it seemed to work. not sure why it isn't here. but we can discuss tomorrow.

Reply:
If it's OK, I'll merge as-is, and then open a separate PR on omegaconf-related typing ignores.

Reply:
yup, works for me. feel free to merge when you are ready.